include/eve/algo/container/detail/soa_storage2.hpp

```diff
@@ -0,0 +1,209 @@
+
+    EVE_FORCEINLINE auto data_aligned()
+    {
+      auto ptrs = kumi::map([]( auto p) { return as_aligned_pointer( p ); }, *this);
+      return kumi::apply(eve::algo::views::zip, ptrs);
+    }
+
+    EVE_FORCEINLINE auto data_aligned() const
+    {
+      auto ptrs = kumi::map([]( auto p) { return as_aligned_pointer( p ); }, *this);
+      return kumi::apply(eve::algo::views::zip, ptrs);
+    }
+
+    EVE_FORCEINLINE auto data()       { return eve::unalign(data_aligned()); }
+    EVE_FORCEINLINE auto data() const { return eve::unalign(data_aligned()); }
+
+    void swap(soa_storage& other) noexcept
+    {
+      data_.swap(other.data_);
+      std::swap(indexes_  , other.indexes_  );
+      std::swap(capacity_ , other.capacity_ );
+    }
+
+    friend void swap(soa_storage& a, soa_storage& b ) noexcept { a.swap(b); }
+
+    //==============================================================================================
+    // Static helpers
+    //==============================================================================================
+    static auto aligned_capacity(std::size_t n) noexcept
+    {
+      // Capacity is the smallest gap possible, ie the one in the largest type
+      constexpr std::size_t largest = max_scalar_size_v<Type>;
+      return eve::align(largest*n, eve::over{64})/largest;
+    }
+
+    static auto byte_size(std::size_t n) noexcept
+    {
+      return  kumi::fold_left ( [n]<typename S>(auto a,S) { return a + realign<S>(n); }
+                              , flat_t{}
+                              , 0ULL
+                              );
+    }
+
+    //==============================================================================================
+    // Tuple interface
+    //==============================================================================================
+    template<std::size_t I>
+    friend decltype(auto) get(soa_storage& s) noexcept
+    {
+      return reinterpret_cast<kumi::element_t<I,Type>*>(s.data_.get() + get<I>(s.indexes_));
+    }
+
+    template<std::size_t I>
+    friend decltype(auto) get(soa_storage const& s) noexcept
+    {
+      return reinterpret_cast<kumi::element_t<I,Type> const*>(s.data_.get() + get<I>(s.indexes_));
+    }
+
+    //==============================================================================================
+    template<typename U> struct as_index { using type = std::size_t; };
+
+    struct aligned_deleter : Allocator
+    {
+      aligned_deleter(Allocator a) : Allocator(a) {}
+      void operator()(byte_t* p) { Allocator::deallocate_aligned(p); }
+    };
+
+    Allocator get_allocator() const { return data_.get_deleter(); }
+
+    //==============================================================================================
+    using data_t    = std::unique_ptr<byte_t, aligned_deleter>;
+    using flat_t    = kumi::result::flatten_all_t<Type>;
+    using indexes_t = kumi::as_tuple_t<flat_t, as_index>;
```

Comment:
std calls them _type. Not that I care.

If things change
data_t - data_ptr_type?

Reply:
Changed to data_ptr_t as this type is not meant to be public